lib/logstash/inputs/beats.rb

```diff
@@ -207,9 +210,229 @@ def invoke(connection, &block)
+      @codec = input.codec.dup
+      
+      @json_transformer = JsonTransform.new(@input)
+      @codec_transformer = CodecTransform.new(@input)
+    end
+
+    def accept
+      @logger.debug("Beats input: waiting from new events from remote host",
+                    :peer => @connection.peer)
+
+      @connection.run { |map, identity_stream| process(map, identity_stream) }
+    end
+
+    def process(map, identity_stream)
+      @logger.debug? && @logger.debug("Beats input: new event received",
+                                      :event_hash => map,
+                                      :identity_stream => identity_stream,
+                                      :peer => @connection.peer)
+
+      # Filebeats uses the `message` key and LSF `line`
+      target_field = @input.target_field_for_codec ? map.delete(@input.target_field_for_codec) : nil
+
+      if target_field.nil?
+        @logger.debug? && @logger.debug("Beats input: not using the codec for this event, can't find the codec target field",
+                                         :target_field_for_codec => @input.target_field_for_codec,
+                                         :event_hash => map)
+
+        event = LogStash::Event.new(map)
+        @json_transformer.transform(event)
+
+        raise InsertingToQueueTakeTooLong if !@queue.offer(event)
+      else
+        @logger.debug? && @logger.debug("Beats input: decoding this event with the codec",
+                                        :target_field_value =>  target_field)
+
+        @codec.accept(CodecCallbackListener.new(target_field,
+                                                  map,
+                                                  identity_stream,
+                                                  @codec_transformer,
+                                                  @queue))
+      end
+    end
+
+    # OOB call to flush the codec buffer, 
+    #
+    # This method is a bit tricky to decide when to be called, in the current case,
+    # this will be call on any exception raised, either is a circuit breaker or the
+    # remote host closed the connection, its better to make sure we clear their
+    # data and create duplicates then losing the data.
+    def flush(&block)
+      @logger.debug? && @logger.debug("Beats input, out of band call for flushing the content of this connection",
+                                      :peer => @connection.peer)
+
+      @codec.flush(&block)
+    end
+  end
+
+  # Base Transform class, expose the plugin decorate method_defined?
+  # apply the tags and make sure we copy the beat hostname into `host`
+  # for backward compatibility.
+  class EventTransformCommon
+    def initialize(input)
+      @input = input
+      @logger = input.logger
+    end
+
+    # Copies the beat.hostname field into the host field unless
+    # the host field is already defined
+    def copy_beat_hostname(event)
+      host = event["beat"] ? event["beat"]["hostname"] : nil
```

Comment:
specially with the new java event, `event["beat"]["hostname"]` should become an anti pattern, to be refactored to `event["[beat][hostname]"]`